sdk/identity/Azure.Identity/src/ProcessRunner.cs

```diff
@@ -0,0 +1,130 @@
+    {
+        private readonly IProcess _process;
+        private readonly TaskCompletionSource<string> _tcs;
+        private readonly CancellationToken _cancellationToken;
+        private readonly CancellationToken _timeoutCancellationToken;
+        private readonly CancellationTokenSource _timeoutCts;
+        private CancellationTokenRegistration _ctRegistration;
+        private CancellationTokenRegistration _timeoutCtRegistration;
+
+        public ProcessRunner(IProcess process, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            _process = process;
+            _cancellationToken = cancellationToken;
+            _tcs = new TaskCompletionSource<string>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            if (timeout.TotalMilliseconds >= 0)
+            {
+                _timeoutCts = new CancellationTokenSource();
+                _timeoutCancellationToken = _timeoutCts.Token;
+                _timeoutCts.CancelAfter(timeout);
+            }
+
+            process.Exited += (o, e) => HandleExit();
+        }
+
+        public Task<string> RunAsync()
+        {
+            StartProcess();
+            return _tcs.Task;
+        }
+
+        public string Run()
+        {
+            StartProcess();
+#pragma warning disable AZC0102 // Do not use GetAwaiter().GetResult().
+            return _tcs.Task.GetAwaiter().GetResult();
+#pragma warning restore AZC0102 // Do not use GetAwaiter().GetResult().
+        }
+
+        private void StartProcess()
+        {
+            if (!TrySetCanceled() && !_tcs.Task.IsCompleted)
+            {
+                _process.Start();
+                _ctRegistration = _cancellationToken.Register(HandleCancel, false);
+                _timeoutCtRegistration = _timeoutCancellationToken.Register(HandleCancel, false);
+            }
+        }
+
+        private void HandleExit()
+        {
+            if (_process.ExitCode == 0)
+            {
+                TrySetResult(_process.StandardOutput.ReadToEnd());
+            }
+            else
+            {
+                TrySetException(new InvalidOperationException(_process.StandardError.ReadToEnd()));
+            }
+        }
+
+        private void HandleCancel()
+        {
+            if (_tcs.Task.IsCompleted)
+            {
+                return;
+            }
+
+            if (!_process.HasExited)
+            {
+                try
+                {
+                    _process.Kill();
+                }
+                catch (Exception ex)
+                {
+                    TrySetException(ex);
+                    return;
+                }
+            }
+
+            TrySetCanceled();
+        }
+
+        private void TrySetResult(string result)
+        {
+            DisposeProcess();
+            _tcs.TrySetResult(result);
+        }
+
+        private bool TrySetCanceled() => TrySetCanceled(_timeoutCancellationToken) || TrySetCanceled(_cancellationToken);
```

Comment:
return value is not used.

Reply:
It is used:
```cs
        private void StartProcess()
        {
            if (!TrySetCanceled() && !_tcs.Task.IsCompleted)
            {
                _process.Start();
                _ctRegistration = _cancellationToken.Register(HandleCancel, false);
                _timeoutCtRegistration = _timeoutCancellationToken.Register(HandleCancel, false);
            }
        }
```

Reply:
Can use `_cancellationToken.ThrowIfCancellationRequested` here?